src/core/mutatus.scala

```diff
@@ -30,25 +33,61 @@ object `package` {
     DatastoreOptions.getDefaultInstance.getService
   )
 
+  /** provides `saveAll` and `deleteAll` methods for collections of case class instances */
+  implicit class DataBatchExt[T <: Product](values: Iterable[T]) {
+
+    /** saves the all case class as a Datastore entity in batch mode */
+    def saveAll()(implicit svc: Service,
+                  encoder: Encoder[T],
+                  dao: Dao[T],
+                  idField: IdField[T]): Map[T, Ref[T]] = {
+      val (batch, refs) =
+        values.foldLeft(svc.readWrite.newBatch() -> mutable.Map.empty[T, Ref[T]]) {
+          case ((b, entityRefs), value) =>
+            val entity = value.buildEntity()
+            b.put(value.buildEntity())
+            b -> entityRefs.updated(value, new Ref[T](entity.getKey))
+        }
+      batch.submit()
+      refs.toMap //batch may supply only those keys which were generated by Datastore, but in mutatus generates keys deterministically based on entity content
+    }
+
+    /** deletes the Datastore entities in batch mode*/
+    def deleteAll()(implicit svc: Service, dao: Dao[T], idField: IdField[T]): Unit = {
+      val _ = values
+        .foldLeft(svc.readWrite.newBatch()) {
+          case (b, value) =>
+            b.delete(idField.idKey(idField.key(value)).newKey(dao.keyFactory))
+            b
+        }
+        .submit()
+    }
+  }
+
   /** provides `save` and `delete` methods on case class instances */
   implicit class DataExt[T <: Product](value: T) {
 
+    private[mutatus] def buildEntity()(implicit encoder: Encoder[T],
+                                       dao: Dao[T],
+                                       idField: IdField[T]): Entity =
+      encoder.encode(value) match {
+        case entityValue: EntityValue =>
+          Entity
+            .newBuilder(
+              idField.idKey(idField.key(value)).newKey(dao.keyFactory),
+              entityValue.get()
+            )
+            .build()
+        case _ => throw MutatusException("Unable to store non entity type")
```

Comment:
Are there cases where `encode` returns something other than an `EntityValue`?

Reply:
Yes, collections are stored as `ListValue`, Option.empty as `NullValue` and primitives as `LongValue`,`StringValue` etc. Also `Ref[T]` is stored as `KeyValue`. However in case of case classes/ sealed traits they are always `EntityValue`